### PR TITLE
Converts parallax to pixel offsets, saves a bunch of cpu time, makes things nicer on clients too

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Right now parallax is like a quarter of SSinput, which is BAD. It's so high mostly because of the animates we need to do, but also due to the cost of setting screen_loc.

image

This sucks. The default step is to reduce the poll rate of the effect, but I don't want to do that because it SUCKS. Sooooo how can we optimize.

Well, if we stop thinking in terms of screen_loc, which is a string (tree shit) and also unanimatable, and start working in pixel offsets, this'd be a way cheaper.

We can make that happen by sticking all our parallax layers on one rock screen object. Then they have relative positions and can be pixel offset (I have stolen this concept wholesale from Ter)

This works unreasonably well, roughly a 65% cost reduction. S good shit.

image

While I'm here...

[uses KEEP_TOGETHER to reduce clientside load, makes the flying animation better.](https://github.com/tgstation/tgstation/commit/52610398e2dc221c80d1f629e9c9f8fb59498977)

We were individually rendering all like fucking 24 480x480 overlays on all 5 parallax layers, which means we had to apply our transform to EACH ONE. This has GOTTA suck shit for clients, so let's... not? Should help.

The existing flying animation makes me depressed. it has some very visible stutter, and jumps around a lot.

We can deal with the starting stutter by avoiding starting a new animation on the layer until the old one is finished. This is what was SUPPOSED to be happening, but because we fired one timer for all the layers, they'd desync and jump in ugly ways.

This means we need to use one timer per layer, which does induce more cost then I'd like. IDK how I feel about this to be honest.

I try and reduce ending weirdness by unscaling time at the end, so different aspects don't slow down at different rates.

Speed on the parallax animation was weird, it'd spike up, then dip down in flight. 
This was because the percieved rate of change from the quad easing was closer to 2x the existing. 
I've handled this by halving the animation time in the loop

Oh also there's no sense calling the update animation proc if we are coming to a stop, and thus have no follow up animation.

## Changelog

:cl: LemonInTheDark
refactor: I have reworked how parallax and its animations (space travel) work. Please report any bugs lads!
/:cl:

### CODE REVIEW

File: code/_onclick/hud/parallax/parallax.dm [hunk 1]
```diff
@@ -11,6 +11,10 @@
 	for(var/atom/movable/screen/plane_master/parallax as anything in get_true_plane_masters(PLANE_SPACE_PARALLAX))
 		parallax.unhide_plane(screenmob)
 
+	if(!C.parallax_rock)
```
Comment (san7890) [inline code/_onclick/hud/parallax/parallax.dm:14]:
```suggestion
	if(isnull(C.parallax_rock))
```

**Review by san7890 — CHANGES_REQUESTED** [review]
way minor nit but it's a nice one

**Comment by san7890** [comment]
also should we have a changelog so players know that parallax was futzed with and to file issue reports on anything weird or does it not matter in this context

File: code/_onclick/hud/parallax/parallax.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/_onclick/hud/parallax/parallax.dm:14]: I TRIED TO WARD AGAINST YOUR EVIL WAYS, BEAST

**Comment by LemonInTheDark** [comment]
> also should we have a changelog so players know that parallax was futzed with and to file issue reports on anything weird or does it not matter in this context

since you think it's worthwhile, sure

**Pull request merged**